source/sayAll.py

```diff
@@ -1,273 +1,289 @@
-		# We've just started speaking this line, so move the cursor there.
-		state.updateObj()
-		updater = obj.makeTextInfo(bookmark)
-		if self.cursor == CURSOR_CARET:
-			updater.updateCaret()
-		if self.cursor != CURSOR_CARET or config.conf["reviewCursor"]["followCaret"]:
-			api.setReviewPosition(updater, isCaret=self.cursor==CURSOR_CARET)
-		winKernel.SetThreadExecutionState(winKernel.ES_SYSTEM_REQUIRED)
-		if self.numBufferedLines == 0:
-			# This was the last line spoken, so move on.
-			self.nextLine()
-		else:
-			self.numBufferedLines -= 1
-
-	def turnPage(self):
-		try:
-			self.reader.obj.turnPage()
-		except RuntimeError:
-			log.debug("No more pages")
-			# No more pages.
-			self.stop()
-			return
-		self.reader = self.reader.obj.makeTextInfo(textInfos.POSITION_FIRST)
-		self.nextLine()
-
-	def finish(self):
-		# There is no more text.
-		# Call stop to clean up, but only after speech completely finishes.
-		# Otherwise, if a different synth is being used for say all,
-		# we might switch synths too early and truncate the final speech.
-		# We do this by putting a CallbackCommand at the start of a new utterance.
-		cb = CallbackCommand(self.stop, name="say-all:stop")
-		_speechWithoutPausesInstance.speakWithoutPauses([
-			EndUtteranceCommand(),
-			cb,
-			EndUtteranceCommand()
-		])
-
-	def stop(self):
-		if not self.reader:
-			return
-		self.reader = None
-		self.trigger.exit()
-		self.trigger = None
-
-	def __del__(self):
-		self.stop()
-
-class SayAllProfileTrigger(config.ProfileTrigger):
-	"""A configuration profile trigger for when say all is in progress.
-	"""
-	spec = "sayAll"
+# A part of NonVisual Desktop Access (NVDA)
+# Copyright (C) 2006-2021 NV Access Limited
+# This file may be used under the terms of the GNU General Public License, version 2 or later.
+# For more details see: https://www.gnu.org/licenses/gpl-2.0.html
+
+from typing import Literal
+import weakref
+import garbageHandler
+import speech
+from logHandler import log
+import config
+import controlTypes
+import api
+import textInfos
+import queueHandler
+import winKernel
+
+from speech.commands import CallbackCommand, EndUtteranceCommand
+
+CURSOR_CARET = 0
+CURSOR_REVIEW = 1
+CURSOR = Literal[CURSOR_REVIEW, CURSOR_CARET]
+
+
+class _SayAllHandler:
+	def __init__(self):
+		self.lastSayAllMode = None
+		self._speechWithoutPausesInstance = None
+		#: The active say all manager.
+		#: This is a weakref because the manager should be allowed to die once say all is complete.
+		self._getActiveSayAll = lambda: None  # noqa: Return None when called like a dead weakref.
+
+	@property
+	def speechWithoutPausesInstance(self) -> 'speech.SpeechWithoutPauses':
+		if self._speechWithoutPausesInstance is None:
+			log.error('attempting to use sayAll without calling initialize first')
+			self.initializeSpeechWithoutPauses()  # TODO: or raise Error?
+			# raise AttributeError('call initializeSpeechWithoutPauses first before using sayAllHandler')
```

Comment:
possibly better to raise an error here - which could require another note in the changelog about calling `initializeSpeechWithoutPauses` before otherwise using `SayAllHandler`